File: php/preorderTraversal.php
<?php

class TreeNode
{
    public $val = null;
    public $left = null;
    public $right = null;

    function __construct($val = 0, $left = null, $right = null)
    {
        $this->val = $val;
        $this->left = $left;
        $this->right = $right;
    }
}

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root
     * @return Integer[]
     */
    function preorderTraversal($root) {
        $orders = [];
        if (is_null($root)) {
            return $orders;
        }

        $orders[] = $root->val;
        if (!is_null($root->left)) {
            $orders = array_merge($orders, $this->preorderTraversal($root->left));
        }
        if (!is_null($root->right)) {
            $orders = array_merge($orders, $this->preorderTraversal($root->right));
        }
        return $orders;
    }
}

var_dump(
    (new Solution())->preorderTraversal(
        (new TreeNode(
            1,
            null,
            (new TreeNode(
                2,
                (new TreeNode(3))
            ))
        ))
    )
);

File: go/preorderTraversal.go
package main

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func preorderTraversal(root *TreeNode) []int {
    var result []int
    var stack [] *TreeNode

    for ;; {
        if root == nil {
            stackSize := len(stack)
            if (stackSize > 0) {
                root = stack[stackSize - 1]
                stack = stack[:stackSize - 1]
            }
        }

        if root != nil {
            result = append(result, root.Val)
            if root.Left != nil {
                if root.Right != nil {
                    stack = append(stack, root.Right)
                }
                root = root.Left
            } else if root.Right != nil {
                root = root.Right
            } else {
                root = nil
            }
        } else {
            break
        }
    }

    return result
}
